src/preamble/02_what_is_a_module_and_a_package.md

```diff
@@ -1 +1,194 @@
 # What is a module and a package?
+
+## TL;DR
+
+A **module** is a unit of code defined by GHC and combined by Cabal to create
+packages. A **package** is either a library (a re-usable collection of modules),
+or an executable (an application that is the combination of one or more modules
+into a binary executable). A test suite is kind of like a package, in the sense
+that it is specified separately at the top-level of a `<project-name>.cabal`
+file, and is built up out of modules. However, it is different in the sense that
+it is usually not exposed and distributed. Broadly speaking, the point of Cabal
+is to define packages.
+
+## What is a module?
+
+At its most basic, a module is a namespace for collecting types, functions, and
+typeclasses (referred to as entities). Modules are a common feature in most
+languages, and the ML family of languages in particular are known for having a
+very robust module system (OCaml has full blown ML modules, and Haskell has
+partial support for them with [backpack](../getting_fancy/10_backpack.md) but
+that's a slightly more advanced topic). Unlike ML modules, Haskell modules are
+not first class (they cannot be passed around as values). Despite this, Haskell
+modules are very feature-full, they can export all or some of their entities,
+and imports of modules can: be qualified with a prefix, explicitly describe a
+few entities to bring into scope, or explicitly hide a few entities and bring
+the rest into scope. If you want to learn more about Haskell modules and their
+syntax, then the
+[haskell2010 report](https://www.haskell.org/onlinereport/haskell2010/haskellch5.html)
+is a good place to start.
+
+The humble module, at its core, is just a namespace, but it provides powerful
+practical capabilities.
+
+Modules can be used as an abstraction layer, by limiting what they export. In
+this way one can create opaque types.
+
+```haskell
+module Money (
+  PositiveDollar,
+  mkPositiveDollar
+) where
+
+newtype PositiveDollar = PositiveDollar { unPositiveDollar :: Int }
+
+mkPositiveDollar :: Int -> Maybe PositiveDollar
+mkPositiveDollar n
+  | n >= 0 = Just . PositiveDollar $ n
+  | otherwise = Nothing
+```
+
+> Note: If you want to learn more about opaque types and their use cases this
+> [resource](https://www.cs.auckland.ac.nz/references/haskell/haskell-intro-html/modules.html)
+> is fantastic
+
+Modules can be used to separate concerns, and split up chunks of code into
+logical and manageable pieces; an 80 line file is much more approachable, and
+easy to take in at a glance, than a 1000 line file.
+
+Module separation can also impact compile times. Each module keeps track of the
+fingerprint (A unique identifier, usually the hash of the contents of the file)
```

Comment:
This is very new. I don't think hash-based recompilation avoidance has been backported to any GHC release yet.